department-of-veterans-affairs/caseflow
# Added validation to avoid ChangeHearingRequestType duplication

Resolves [CASEFLOW-1512](https://vajira.max.gov/browse/CASEFLOW-1512)

### Description
In #15805 we implemented the functionality to change hearing request type by creating and immediately closing ChangeHearingRequestTypeTask.

We've had a couple of bat team tickets lately ([one](https://dsva.slack.com/archives/CHX8FMP28/p1619800479142300), [two](https://dsva.slack.com/archives/CHX8FMP28/p1620665263264000)) where tasks were created even though there was no actual change being requested (i.e. the hearing request type was virtual, and the params requested a change to virtual).

Stop this from happening by not creating a new ChangeHearingRequestTypeTask in response to this type of request unless the new hearing request type is different from what's currently set.

### Acceptance Criteria
- [ ] When converting an Appeal only one `ChangeHearingRequestTypeTask` is created.

### Testing Plan

1. First test on `master` so you can reproduce
2. Shadow BVASYELLOW
3. Go to http://localhost:3000/hearings/schedule
4. Click on "Schedule Veterans"
5. Select St. Petersburg, FL RO
6. Click on "AMA Veterans Waiting"
7. Click on any Veteran
8. Open one tab/window where there's a video hearing request, and you can see the conversion dropdown.
9. Open another tab/window to the same page
10. In the first window, convert the hearing request to virtual.
11. In the second window, convert the hearing request to virtual.

A couple of things should have happened:

- The appeals show page should error out.  ( http://localhost:3000/queue/appeals/{uuid} )
- The `Appeal` should have 2 `ChangeHearingRequestTypeTask`
- Open the console `make c`
- Find the appeal:

  ```ruby
  ap = Appeal.find_by(uuid: uuid)
  ap.treee

  # This print to the console an extra ChangeHearingRequestTypeTask created.
  # If we delete the last one on the list the Appeal show page should now be accessible
  ```

12. Now checkout the working branch `git checkout omar/CASEFLOW-1512-change-hearing-request-type-validation`
13. Repeat all the steps above on and only one `ChangeHearingRequestTypeTask` is created.
